android/KMEA/app/src/main/java/com/tavultesoft/kmea/KMHardwareKeyboardInterpreter.java

```diff
@@ -114,18 +116,70 @@ public KMHardwareKeyboardInterpreter(Context context, KMManager.KeyboardType key
 
   @Override
   public boolean onKeyDown(int keyCode, KeyEvent event) {
+    // Note: Keep this table in sync with kmwosk.js osk.modifierCodes
+    final HashMap<String, Integer> modifierCodes = new HashMap<String, Integer>() {{
+      put("LCTRL", 0x0001);
+      put("RCTRL", 0x0002);
+      put("LALT", 0x0004);
+      put("RALT", 0x0008);
+      put("SHIFT", 0x0010);
+      put("CTRL", 0x0020);
+      put("ALT", 0x0040);
+      put("CAPS", 0x0100);
+      put("NO_CAPS", 0x0200);
+      put("NUM_LOCK", 0x0400);
+      put("NO_NUM_LOCK", 0x0800);
+      put("SCROLL_LOCK", 0x1000);
+      put("NO_SCROLL_LOCK", 0x2000);
+      put("VIRTUAL_KEY", 0x4000);
+    }};
+
     if (keyCode > 84 || keyCode < 0) {
       // The key is outside the range of keys we understand
       return false;
     }
 
+    boolean isChiral = false;
+    if (this.keyboardType == KMManager.KeyboardType.KEYBOARD_TYPE_INAPP) {
+      isChiral = KMManager.InAppKeyboard.getChirality();
+    } else if (this.keyboardType == KMManager.KeyboardType.KEYBOARD_TYPE_SYSTEM) {
+      isChiral = KMManager.SystemKeyboard.getChirality();
```

Comment:
Not that it belongs in this PR, but I remember there being a LOT of spots in the Android code with this pattern.  Maybe we should refactor things eventually with a `KMManager.getActiveKeyboard()` function to consolidate this constant condition check?

Reply:
There's 3 patterns that are "same same but different"

```java
if (keyboard == KeyboardType.KEYBOARD_TYPE_INAPP) {
  ...
} else if (keyboard == KeyboardType.KEYBOARD_TYPE_SYSTEM) {
  ...
}
```
and
```java
if (InAppKeyboard != null) {
  ...
else if (SystemKeyboard != null) {
  ...
}
```
and
```java
if (InAppKeyboard != null) {
  ...
}
if (SystemKeyboard != null) {
  ...
}

I can have `KMManager.getActiveKeyboard()` cover one, and leave the refactor as a TODO for another day.

Reply:
Let's leave the refactor to a specific refactoring sprint.